packages/assets-controllers/src/CurrencyRateController.ts

```diff
@@ -163,69 +128,33 @@ export class CurrencyRateController extends BaseControllerV2<
    * @param currentCurrency - ISO 4217 currency code.
    */
   async setCurrentCurrency(currentCurrency: string) {
-    this.update((state) => {
-      state.pendingCurrentCurrency = currentCurrency;
-    });
-    await this.updateExchangeRate();
-  }
-
-  /**
-   * Sets a new native currency.
-   *
-   * @param symbol - Symbol for the base asset.
-   */
-  async setNativeCurrency(symbol: string) {
-    this.update((state) => {
-      state.pendingNativeCurrency = symbol;
-    });
-    await this.updateExchangeRate();
-  }
-
-  private stopPolling() {
-    if (this.intervalId) {
-      clearInterval(this.intervalId);
+    const releaseLock = await this.mutex.acquire();
+    const nativeCurrencies = Object.keys(this.state.currencyRates);
+    try {
+      this.update(() => {
+        return {
+          ...defaultState,
+          currentCurrency,
+        };
+      });
+    } finally {
+      releaseLock();
     }
+    nativeCurrencies.forEach(this.updateExchangeRate.bind(this));
```

Comment:
This ends up being a lot of state updates right? Would it be be better to have updateExchangeRate return the updated rates, and we roll all the `this.update` into 1?

Also, What is the lock for? I presume its to prevent you from updating the current currency before the first update is complete, but then shouldnt the `nativeCurrencies.forEach(this.updateExchangeRate.bind(this));` be inside the lock? (and awaited / promise.all'd)?

Reply:
> Would it be be better to have updateExchangeRate return the updated rates, and we roll all the this.update into 1

Do you mean replacing the `updateExchangeRate` calls with a `fetchExchangeRate` call here in the loop and then updating manually... We might want to refactor these methods a bit if we wanted to take this approach. Not sure how much performance it buys us.